File: evals/fixtures/java-service/src/main/java/checkout/CheckoutServer.java
// Eval fixture: deliberately uninstrumented Java HTTP service (see
// evals/fixtures/README.md for the contract). It is a Spring Boot Web
// application whose embedded Tomcat plus Spring WebMVC stack the scenario's
// zero-code Java agent auto-instruments, so a SERVER span is produced for
// GET /checkout. It serves GET /checkout, performs one outbound HTTP call to
// DOWNSTREAM_URL while handling it, listens on PORT (default 8080, wired
// through application.properties), and uses only obviously synthetic data
// (user@example.test, TEST-0001). No instrumentation dependencies of its
// own beyond Spring Boot Web: the scenario attaches the agent at run time.
package checkout;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

// Spring configuration classes (implied by @SpringBootApplication) may not be
// final, so this class is intentionally left non-final.
@SpringBootApplication
@RestController
public class CheckoutServer {

    // A single reused client keeps the outbound behavior to exactly one GET
    // per handled request, matching the fixture contract.
    private final HttpClient client = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build();

    private final String downstreamUrl =
            env("DOWNSTREAM_URL", "http://localhost:9090/inventory");

    public static void main(String[] args) {
        SpringApplication.run(CheckoutServer.class, args);
    }

    @GetMapping(path = "/checkout", produces = MediaType.APPLICATION_JSON_VALUE)
    public Map<String, Object> checkout() {
        HttpRequest request = HttpRequest.newBuilder(URI.create(downstreamUrl))
                .timeout(Duration.ofSeconds(5))
                .GET()
                .build();
        String inventory;
        try {
            HttpResponse<String> response =
                    client.send(request, HttpResponse.BodyHandlers.ofString());
            inventory = response.body();
        } catch (Exception e) {
            throw new InventoryUnavailableException(e);
        }

        return Map.of(
                "order_id", "TEST-0001",
                "customer_email", "user@example.test",
                "status", "confirmed",
                "inventory", inventory);
    }

    private static String env(String name, String fallback) {
        String value = System.getenv(name);
        return value == null || value.isEmpty() ? fallback : value;
    }

    // Surfaces a failed inventory lookup as a 502 without pulling in extra
    // Spring types; the default error handling maps the annotation.
    @org.springframework.web.bind.annotation.ResponseStatus(
            org.springframework.http.HttpStatus.BAD_GATEWAY)
    private static final class InventoryUnavailableException extends RuntimeException {
        InventoryUnavailableException(Throwable cause) {
            super("inventory unavailable", cause);
        }
    }
}

File: evals/custom/fixtures/scala-service/src/main/scala/checkout/CheckoutServer.scala
// Eval fixture: deliberately uninstrumented Scala HTTP service (see
// evals/fixtures/README.md for the contract). It serves GET /checkout,
// performs one outbound HTTP call to DOWNSTREAM_URL while handling it,
// listens on PORT (default 8080), and uses only obviously synthetic data
// (user@example.test, TEST-0001). JDK classes only: no dependencies.
package checkout

import com.sun.net.httpserver.{HttpExchange, HttpServer}

import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.net.{InetSocketAddress, URI}
import java.nio.charset.StandardCharsets
import java.time.Duration
import scala.util.{Failure, Success, Try}

object CheckoutServer {

  def main(args: Array[String]): Unit = {
    val port = sys.env.getOrElse("PORT", "8080").toInt
    val downstreamUrl = sys.env.getOrElse("DOWNSTREAM_URL", "http://localhost:9090/inventory")
    val client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).build()

    val server = HttpServer.create(new InetSocketAddress(port), 0)
    server.createContext("/checkout", (exchange: HttpExchange) => handleCheckout(exchange, client, downstreamUrl))
    server.start()
    println(s"scala-service listening on :$port")
  }

  private def handleCheckout(exchange: HttpExchange, client: HttpClient, downstreamUrl: String): Unit = {
    if (exchange.getRequestMethod != "GET") {
      respond(exchange, 405, """{"error":"method not allowed"}""")
      return
    }

    val request = HttpRequest
      .newBuilder(URI.create(downstreamUrl))
      .timeout(Duration.ofSeconds(5))
      .GET()
      .build()

    Try(client.send(request, HttpResponse.BodyHandlers.ofString())) match {
      case Failure(error) =>
        System.err.println(s"inventory lookup failed: $error")
        respond(exchange, 502, """{"error":"inventory unavailable"}""")
      case Success(response) =>
        println("checkout completed order.id=TEST-0001 customer.email=user@example.test")
        val inventory = quote(response.body())
        respond(
          exchange,
          200,
          s"""{"order_id":"TEST-0001","customer_email":"user@example.test","status":"confirmed","inventory":$inventory}"""
        )
    }
  }

  private def respond(exchange: HttpExchange, status: Int, body: String): Unit = {
    val bytes = body.getBytes(StandardCharsets.UTF_8)
    exchange.getResponseHeaders.set("Content-Type", "application/json")
    exchange.sendResponseHeaders(status, bytes.length.toLong)
    val out = exchange.getResponseBody
    try out.write(bytes)
    finally out.close()
  }

  private def quote(raw: String): String =
    "\"" + raw.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n") + "\""
}
